Sources/Views/MessageLabel.swift

```diff
@@ -22,417 +22,491 @@
+        return NSMutableAttributedString(attributedString: attributedString)
+    }
+    
+    func addAttribute(key: NSAttributedStringKey, value: AnyObject, range: NSRange? = nil) {
+        let range = range ?? NSRange(location: 0, length: length)
+        addAttribute(key, value: value, range: range)
+    }
+    
+}
 
-    open var enabledDetectors: [DetectorType] = [.phoneNumber, .address, .date, .url]
+open class MessageLabel: UIView, UIGestureRecognizerDelegate {
 
-    open override var attributedText: NSAttributedString? {
-        didSet {
-            guard attributedText != oldValue else { return }
-            setTextStorage()
+    // MARK: - Properties [Public]
+    
+    public weak var delegate: MessageLabelDelegate?
+    
+    public var attributedText: NSAttributedString? {
+        get {
+            return textStorage.attributedString
+        }
+        set {
+            let string = newValue ?? NSAttributedString(string: "")
+            textStorage.setAttributedString(string)
+            
+            setStorageByParsing(text: string) {
+                if !self.isConfiguring { self.setNeedsDisplay() }
+            }
+            
+            if !isConfiguring { setNeedsDisplay() }
         }
     }
-
-    open override var text: String? {
-        didSet {
-            guard text != oldValue else { return }
-            setTextStorage()
+    
+    public var text: String? {
+        get {
+            return textStorage.string
+        }
+        set {
+            let attributedString = NSAttributedString(string: newValue ?? "")
+            let string = addLocalAttributes(to: attributedString)
+            textStorage.setAttributedString(string)
+            
+            setStorageByParsing(text: string) {
+                if !self.isConfiguring { self.setNeedsDisplay() }
+            }
+            
+            if !isConfiguring { setNeedsDisplay() }
         }
     }
-
-    open override var font: UIFont! {
+    
+    public var font: UIFont = UIFont.systemFont(ofSize: 10.0) {
         didSet {
-            guard font != oldValue else { return }
-            guard let attributedText = attributedText else { return }
-            textStorage.setAttributedString(attributedText)
-            setNeedsDisplay()
+            textStorage.addAttribute(key: .font, value: font)
+            updateDetectorAttributes()
+            if !isConfiguring {
+                setNeedsDisplay()
+            }
         }
     }
-
-    open override var textColor: UIColor! {
+    
+    public var textColor: UIColor = .darkText {
         didSet {
-            guard textColor != oldValue else { return }
-            guard let attributedText = attributedText else { return }
-            textStorage.setAttributedString(attributedText)
-            setNeedsDisplay()
+            textStorage.addAttribute(key: .foregroundColor, value: textColor)
+            updateDetectorAttributes()
```

Comment:
Do we need to update the detectors when we change the text color?

Reply:
Yeah :( Working on this but for example if the user is using a `UIColor.red` for the textColor of detected attribtues, when we change the font here it will override the detected attributes. So we have to reapply them over it. This method has an optional parameter `key` that allows us to update the detectors for just that key. I _may_ be able to use that here but not sure

Reply:
Ah I see